rsocket-core/src/main/java/io/rsocket/resume/InMemoryResumeStore.java

```diff
@@ -0,0 +1,113 @@
+package io.rsocket.resume;
+
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoProcessor;
+import reactor.util.concurrent.Queues;
+
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class InMemoryResumeStore implements ResumeStore {
+  private static final Logger logger = LoggerFactory.getLogger(InMemoryResumeStore.class);
+
+  private final MonoProcessor<Void> disposed = MonoProcessor.create();
+  private final AtomicLong position = new AtomicLong();
+  private final AtomicLong impliedPosition = new AtomicLong();
+  private volatile AtomicInteger cachedFramesSize = new AtomicInteger();
+  private final Queue<ByteBuf> cachedFrames;
+  private final String tag;
+  private final int cacheLimit;
+
+  public InMemoryResumeStore(String tag, int cacheLimit) {
+    this.tag = tag;
+    this.cacheLimit = cacheLimit;
+    this.cachedFrames = cachedFramesQueue(cacheLimit);
+  }
+
+  @Override
+  public void saveFrame(ByteBuf frame) {
+    cachedFramesSize.incrementAndGet();
+    cachedFrames.offer(frame.retain());
+    if (cachedFramesSize.get() == cacheLimit) {
+      releaseFrame();
+    }
+  }
+
+
+  @Override
+  public void releaseFrames(long remoteImpliedPos) {
+    long pos = position.get();
+    logger.info("{} Removing frames for local: {}, remote implied: {}", tag, pos, remoteImpliedPos);
+    long removeCount = Math.max(0, remoteImpliedPos - pos);
+    long processedCount = 0;
+    while (processedCount < removeCount) {
+      releaseFrame();
+      processedCount++;
+    }
+    logger.info("{} Removed frames. Current size: {}", tag, cachedFramesSize.get());
+  }
+
+  private void releaseFrame() {
+    ByteBuf content = cachedFrames.poll();
+    cachedFramesSize.decrementAndGet();
+    content.release();
+    position.incrementAndGet();
+  }
+
+  /*must be subscribed at most once on resumeStart -> resumeFrames cycle*/
+  @Override
+  public Flux<ByteBuf> resumeStream() {
+    return Flux.create(s -> {
+      int size = cachedFramesSize.get();
+      logger.info("{} Resuming stream size: {}",tag, size);
+      /*spsc queue has no iterator - iterating by consuming*/
+      for (int i = 0; i < size; i++) {
+        ByteBuf frame = cachedFrames.poll();
```

Comment:
Should this be destructive in case this resume get disconnected?

Reply:
Most apps dealing with `reactor` and refcounted objects have `Refcounted.release()` in `Hooks.onNextDropped` - otherwise cancelled streams would lead to memory leaks. The idea here is to retain frames for future resumes, and assume calling code will release accordingly (either transport or `Hooks.onNextDropped`)